cluster/querier/querier.go

```diff
@@ -0,0 +1,94 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+// Package querier provides abstractions to keep track of and manage querier nodes.
+package querier
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// QuerierManager keeps track of registered querier nodes and allows one to notify all of them
+// of class/tenant data updates.
+type QuerierManager struct {
+	mu                 sync.Mutex
+	registeredQueriers map[*Querier]struct{}
```

Comment:
If you're using the lock to sync access to the map only, you can just use an `async.Map` instead to simplify the code.

Reply:
ALso the sync map provides range https://pkg.go.dev/sync#Map.Range which allows us to delete a key while iterating the map elsewhere

Reply:
Good idea, thanks!